cmake/toolchain/llvm/generic.cmake

```diff
@@ -17,9 +17,11 @@ if("${ARCH}" STREQUAL "arm")
   set(CMAKE_EXE_LINKER_FLAGS_INIT "--specs=nosys.specs")
 elseif("${ARCH}" STREQUAL "x86")
   if(CONFIG_64BIT)
-    set(triple x86_64-pc-none-elf)
+    # The LLVM built-in linker lld is a generic driver, needs to use triple os filed
+    # to distinguish which linker will be used.
+    set(triple x86_64-pc-linux-elf)
```

Comment:
Does it force lld to generate an elf binary on Windows?

Reply:
yes, this is a target name for cross compiler, but on windows, the older name "i686-pc-none-elf" couldn't be recognized and used to find a valid linker, we need to use "i686-pc-linux-elf".

Reply:
where is this coming from? Why does llvm on windows require a triple with linux in it? Is this in LLVM?

Reply:
@nashif as I understood, we need to build a zephyr elf file which could be executed on linux, it's a cross compilation from windows to linux, and llvm on windows has four linker based on os type: ld.lld (Unix), ld64.lld (macOS), lld-link (Windows), wasm-ld (WebAssembly), so I added linux in triple name, then it will use this ld.lld.

Reply:
`i686-pc-none-elf` refers to the x86 (i686) ELF bare metal target, while `x86_64-pc-linux-elf` refers to the x86 (i686) ELF Linux target.

I do not know how LLVM internally handles this; but, at least in terms of GNU toolchains, they are not the same thing and we must use `i686-pc-none-elf` to ensure correct behaviour.

If the Windows LLVM distribution does not support `i686-pc-none-elf` (i.e. bare metal target) while the Linux one does, we should be filing an LLVM-side bug instead to make it support that -- unless, of course, there is a good reason why it should not support bare metal targets on Windows hosts (I really cannot see any).

Reply:
By the way, have you tried `i686-none-elf` or `i686-unknown-none-elf` by chance? Because `pc` here is not so technically correct (although it just refers to quite meaningless vendor name and does not matter for GNU toolchains, LLVM might have different regex rules handling this differently).

Reply:
@stephanosio both on windows and linux, llvm doesn't support triple name i686-pc-none-elf or i686-unknown-none-elf.

I tested to build an empty file in linux using  i686-unknown-none-elf with command as below, it won't use the llvm built-in linker even I added option "-fuse-ld=lld".

command: ./clang dummy.c -o dummy -target i686-unknown-none-elf -fuse-ld=lld -v
**"/usr/lib64/ccache/gcc" -fuse-ld=lld -m32 -o dummy /tmp/dummy-28c5c8.o
gcc: error: unrecognized command line option ‘-fuse-ld=lld’; did you mean ‘-fuse-ld=bfd’?**

**but it worked well with i686-unknown-linux-elf, no error reported.**
**command: ./clang dummy.c -o dummy -target i686-unknown-linux-elf -fuse-ld=lld -v**
/llvm/llvm-project/build_2/bin/./ld.lld" --hash-style=gnu --eh-frame-hdr -m elf_i386 -dynamic-linker /lib/ld-linux.so.2 -o dummy /lib/crt1.o /lib/crti.o /usr/lib/gcc/x86_64-redhat-linux/8/32/crtbegin.o -L/usr/lib/gcc/x86_64-redhat-linux/8/32 -L/home/chenpeng/llvm/llvm-project/build_2/bin/../lib -L/lib -L/usr/lib /tmp/dummy-5880f9.o -lgcc --as-needed -lgcc_s --no-as-needed -lc -lgcc --as-needed -lgcc_s --no-as-needed /usr/lib/gcc/x86_64-redhat-linux/8/32/crtend.o /lib/crtn.o

Reply:
@chen-png That log doesn't look right. It looks like you haven't installed Clang on your system and your "clang" is calling gcc front-end instead?

Both `i686-pc-none-elf` and `i686-none-elf` as well as `x86_64-pc-none-elf` and `x86_64-none-elf` work for me:
```
$ clang dummy.c -o dummy -target x86_64-none-elf -fuse-ld=lld -nostdlib
ld.lld: warning: cannot find entry symbol _start; defaulting to 0x12A8
```
```
$ clang dummy.c -o dummy -target i686-none-elf -fuse-ld=lld -nostdlib
ld.lld: warning: cannot find entry symbol _start; defaulting to 0x11AC
```

Please make sure that both Clang and LLVM LLD are installed on your system.

Reply:
I installed LLVM 10.0.0 on Windows (same version as the one I have on my Linux machine) and tested this.

I was able to reproduce `clang: error: unable to execute command: program not executable` on Windows host with non-Linux targets, but that is only because Clang internally calls the GCC front-end (supposedly to invoke binutils in order to link the targets that it thinks are not supported by LLD) and I did not have GCC installed/available in the PATH. **Note that the same behaviour is observed on Linux hosts as well, except that GCC is installed there, so no such error message is displayed.**

In theory, if we have gcc/binutils that supports the `x86_64-none-elf` target installed on the Windows host, it will work just fine (note that MSYS2/MinGW-w64 GCC does not support `x86_64-none-elf` target, so using that is not an option; this will not be a problem if we distribute Windows version of Zephyr SDK though, since it supports `*-none-elf` targets).

Now the question is:
1. Does LLD not support bare metal targets (e.g. `*-*-elf`)? (i.e. should Clang be calling GCC front-end/binutils for these targets?)
1. If LLD supports bare metal targets, is there a way to force Clang to directly invoke `ld.lld` for these targets as it does for the Linux targets?

The main problem here is that, on both Linux and Windows, Clang is calling GCC front-end with `-fuse-ld=lld`, and the GCC front-end calls the binutils linker (`collect2`), supposedly ignoring `-fuse-ld=lld` -- at this point, we might as well be just using `-fuse-ld=bfd`.

Reply:
cc @tejlmand 
It looks like there is something funny going on with the invocation of LLD, or rather lack thereof.

Reply:
@stephanosio yes, actually I used a cross compiler i486-elf to complete the link process on windows, rather than the llvm built-in linker.

Reply:
> Both `i686-pc-none-elf` and `i686-none-elf` as well as `x86_64-pc-none-elf` and `x86_64-none-elf` work for me:

it seems that it's related to gcc version, I found a comment "**gcc 8.x doesn't support -fuse-ld=lld option at all, it has been only added in http://gcc.gnu.org/r265940 for gcc 9**"
then I installed a gcc 9.4 on my fedora machine, it worked with x86_64-none-elf.

"/usr/lib64/ccache/gcc" -nostdlib -fuse-ld=lld -m32 -o dummy /tmp/dummy-675e45.o
ld.lld: warning: cannot find entry symbol _start; defaulting to 0x401120

Reply:
I used gcc 8.3 before.